### DISCUSSION TITLE
소수 파싱 방법에 대한 논의

### DISCUSSION THREAD

**Discussion by sindydwns** [CubeRT7/minirt#10] [General]
현재 문자열 `""`, `"."`, `".1"` 등이 정상적으로 파싱되고 있습니다.
이 내용은 버그가 아니라 기능으로 개발되었으며 다음과 같은 내용도 정상적으로 파싱되도록 의도하고 있습니다.
`sp  0,0,0  1  0,0,0`
`sp  0,0,  1  0,0,0`
`sp  ,,  1  0,0,0`
`sp  .,.,.  1  0,0,0`
모두 같은 결과로 파싱되고 있으며 이 파싱을 허용할지에 대한 논의 입니다.

**Comment by iijung** [comment]
**⭕  정상적인 문자열이라고 생각합니다.**
서브젝트에서 소수에 대한 기준이 정해져있지 않기 때문에 문자열을 float으로 변환하는 strtof의 문법을 참고했으며 표준 함수와 출력 결과가 일치합니다. 

## strtof 테스트
- 아래의 코드로 strtof의 `""`,` "."`, `".1"`을 테스트해보면 각각 `0.0`, `0.0`, `0.1`으로 정상 파싱되는 것을 확인할 수 있습니다.
``` 
#include <stdio.h>
#include <stdlib.h>

int	main(int argc, char **argv)
{
	char *remain;

	printf("strtof[ %f ]\n", strtof(argv[1], &remain));
	printf("remain[ %s ]\n", remain);
	perror("");
	return (0);
}
```

# 참고
- [IBM docs - strtod()](https://www.ibm.com/docs/ko/i/7.4?topic=lf-strtod-strtof-strtold-convert-character-string-double-float-long-double)

**Comment by sindydwns** [comment]
최종사용자의 사용성을 고려하여 프로그래밍을 한다면 위 기능은 제공하지 않는 것이 좋습니다.
`0.1` 을 `.1` 로 생략하거나 `0,0,0`을 `,,` 으로 생략할 수 있지만 이런 편의를 제공하는 것은 오히려 사용자의 실수를 장려하는 것이 될 수 있습니다.

예를 들어,
`sp  0,0,20  20  0,0,0` 을 입력하려는데 잘못 입력하여 `sp  0,0, 20  0,0,0` 으로 입력하였다면 오류라고 출력하여 사용자의 실수를 최대한 방지해야 합니다. 사용자가 충분히 실수한만한 것 까지 임의로 해석하여 파싱하는 것은 자바스크립트 언어에서 흔히 볼 수 있는 위험한 행동입니다.
![위험한 행동](https://c17an.netlify.app/static/288cb1fa6f64fe39f00046c4e0eae297/0b533/meme.png)

그렇기 때문에, 현재의 파싱은 편리하지만 동시에 위험한 구현이므로 수정되어야 합니다.

**Reply by iijung** [reply]
그렇다면 strtof와 동일하게 구현한 ft_strtof가 아닌 다른 함수를 작성하거나 strtof 호출 전/후로 문자열이 유효한지 확인하는 로직이 추가되어야할 것 같습니다.

이 변경사항을 적용하기 위해서는 
- 사용자 입장에서는 .1 과 같이 작성할 수 있는 편의성이 없어지게 되고
- 개발자 입장에서는 문자열 확인을 위한 코드가 추가 작성해야헤서 가독성이 떨어질까 염려됩니다. 

개인적으로는 현재 상황이 가장 적절한 구현이라고 생각되므로, 이 부분에 대한 추가 구현 사항은 @sindydwns 님이 수정해주셨으면 좋겠습니다

**Reply by sindydwns** [reply]
저는 아무런 불만이 없습니다. 기존 개발 사항을 유지합시다.

**Comment by iijung** [comment]
여기에서 논의하지 못한 많은 이야기로 인해 5d728a3 에서 해당 요구사항을 반영했습니다.

- `digits.digits` 또는 `.digits` 형식에 위반되면 오류로 처리했습니다.
